File: Java/239_SlidingWindowMaximum.java
public class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int len = nums.length;
        if (len == 0) return nums;
        int[] res = new int[len - k + 1];
        
        Deque<Integer> dq = new LinkedList<Integer>();
        for (int i = 0; i < len; i++) {
        	while (!dq.isEmpty() && dq.peekLast() < nums[i])
        		dq.pollLast();
        	
        	dq.offer(nums[i]);
        	
        	if (i >= k - 1) {
        		res[i - k + 1] = dq.peek();
        		if (nums[i - k + 1] == dq.peek())
        			dq.pollFirst();
        	}
        }
    	return res;
    }
}


public class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int len = nums.length;
        if (len == 0) return nums;
        int[] res = new int[len - k + 1];
        
        Deque<Integer> dq = new LinkedList<Integer>();
        for (int i = 0; i < len; i++) {
        	if (!dq.isEmpty() && dq.peek() < i - k + 1)
        		dq.poll();
        	
        	while (!dq.isEmpty() && nums[i] >= nums[dq.peekLast()])
        		dq.pollLast();
        	
        	dq.offer(i);
        	if (i - k + 1 >= 0) 
        		res[i - k + 1] = nums[dq.peek()];
        }
        return res;
    }
}

File: Cpp/239_SlidingWindowMaximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int> &nums, int k) {
        int len = nums.size();
        if (len == 0) return nums;
        vector<int> res(len - k + 1);
        deque<int> dq;
        
        for (int i = 0; i < len; ++i) {
            if (!dq.empty() && dq[0] < i - k + 1)
                dq.pop_front();
            
            while (!dq.empty() && nums[i] > nums[dq.back()])
                dq.pop_back();
            
            dq.push_back(i);
            
            if (i - k + 1 >= 0)
                res[i - k + 1] = nums[dq[0]];
        }
        return res;
    }
};
